cmd/controller-manager/main.go

```diff
@@ -144,6 +213,144 @@ func main() {
+		}
+	}
+	if err != nil {
+		setupLog.Error(err, "unable to create WDS client for leader election after retries")
+		os.Exit(1)
+	}
+
+	// Check/create namespace for leader election with retry for transient errors
+	var namespaceErr error
+	for i := 0; i < 3; i++ {
+		_, namespaceErr = wdsClient.CoreV1().Namespaces().Get(ctx, lockNamespace, metav1.GetOptions{})
+		if namespaceErr == nil {
+			break
+		}
+		if errors.IsNotFound(namespaceErr) {
+			setupLog.Info("Creating namespace for leader election", "namespace", lockNamespace)
+			_, createErr := wdsClient.CoreV1().Namespaces().Create(ctx, &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: lockNamespace,
+				},
+			}, metav1.CreateOptions{FieldManager: fieldManagerName})
+			if createErr == nil || errors.IsAlreadyExists(createErr) {
+				// AlreadyExists is treated as success - another replica or previous run created it
+				namespaceErr = nil
+				break
+			}
+			if i < 2 {
+				setupLog.Info("Failed to create namespace, retrying", "attempt", i+1, "error", createErr)
+				time.Sleep(time.Duration(i+1) * time.Second)
+			} else {
+				namespaceErr = createErr
+			}
+		} else {
+			if i < 2 && isTransientError(namespaceErr) {
+				setupLog.Info("Failed to check namespace due to transient error, retrying", "attempt", i+1, "error", namespaceErr)
+				time.Sleep(time.Duration(i+1) * time.Second)
+			} else if i < 2 {
+				setupLog.Info("Failed to check namespace, retrying", "attempt", i+1, "error", namespaceErr)
+				time.Sleep(time.Duration(i+1) * time.Second)
+			}
+		}
+	}
+	if namespaceErr != nil {
+		setupLog.Error(namespaceErr, "unable to check/create namespace for leader election after retries")
+		os.Exit(1)
+	}
+
+	// Create resource lock for leader election
+	lock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      lockName,
+			Namespace: lockNamespace,
+		},
+		Client: wdsClient.CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: getUniqueIdentity(),
+		},
+	}
+
+	// Configure leader election
+	// These durations follow the standard Kubernetes leader election patterns:
+	// - LeaseDuration: 60s provides more stability and reduces API server load
+	// - RenewDeadline: 40s allows for timely renewal (2/3 of lease duration)
+	// - RetryPeriod: 5s reduces retry frequency for better performance
+	lec := leaderelection.LeaderElectionConfig{
+		Lock:            lock,
+		ReleaseOnCancel: false, // Set to false as the shutdown behavior constraint is not met
+		LeaseDuration:   60 * time.Second,
+		RenewDeadline:   40 * time.Second,
+		RetryPeriod:     5 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				setupLog.Info("Acquired leadership, starting controllers and workers")
+				startControllersDirectly(ctx, setupLog, wdsRestConfig, itsRestConfig, wdsName, itsName, allowedGroupsSet, ctlrsToStart, wdsClientMetrics, itsClientMetrics)
```

Comment:
This will lead to a duplicate call on `kubernetes.NewForConfig(wdsRestConfig)` (inside binding.NewController called from startControllersDirectly). It would be marginally better to have only one call on `kubernetes.NewForConfig` (and this will help prepare for unit testing of these controllers). Can be addressed in follow-on.